### ISSUE TITLE
Support browser history

### ISSUE THREAD

**Issue by alcinnz** [alcinnz/Odysseus#75]
This is currently in progress under the `features/history+default` branch.

The experience I'm going for is a simple and easily skimmable webpage and to link to it from the newtab page and the application menu. At the same time there's a number of enhancements over the MVP I want which I'll track in a seperate issue.

---

Technically this is being implemented as a "trait" which writes to new database tables and an internal webpage (rendered via my custom "Prosody" Django-like templating language). Visits to this or any other internal page should not be tracked by browser history.

**Comment by alcinnz** [comment]
It's worth noting that this is a nice and minimal testcase for UI infrastructure I'll use elsewhere.

**Comment by alcinnz** [comment]
Currently I've got this mostly implemented, but it's segfaulting when I attempt to visit odysseus:history.

The stacktrace reported by GDB is a little unclear thanks to Vala coroutines, but it appears to be crashing upon reporting a syntax error. I don't know what else would give this stacktrace, but my test for it still works.

**Comment by alcinnz** [comment]
Now I've confirmed that it's the error reporting code for invalid templating syntax that is causing the crash, and have learnt the general area in which my syntax is incorrect. Yes, it's quite meta and I'm also having trouble with the syntax I implemented myself.

I learnt this by adding debugging statements and commenting out some of the error reporting code.

**Comment by alcinnz** [comment]
I've just found the problem with the crash and committed a fix.

The issue was that the parser for my templating language reported on the wrong line (didn't bother to investigate that too much as it's minor and only serves myself and any other eventual contributors). Then when it rendered the error report, it segfaulted upon not finding the error on that line.

So I fixed this by scanning forward through the text for the correct line to report the error on.

**Comment by alcinnz** [comment]
And now I've fixed that syntax error, but it looks as if there's no history for the page to show right now.

So am I not collecting it properly?

**Comment by alcinnz** [comment]
I've now enabled history tracking (and fixed some integration errors), but now the history page is crashing again. But this time I've got a better error message.

So I know that if I don't have any browser history to show, the page works but is useless. But if I (presumably) do have history to show, it crashes with a statement that I'm trying to output more HTML after I indicated that I'm done. So this problem probably has to do with the integration between the templating and SQLite.

**Comment by alcinnz** [comment]
I've just fixed and committed that crash (got in trouble with some deprecated syntax trigging an incorrect order of execution).

But now the data shown on the [2018-03-10-history.txt](page) is totally wrong. It shows the dates (which are totally wrong), but no other info.

**Comment by alcinnz** [comment]
Just checked the database, and while it can be definitely be improved the browser history present in the database is much better than what is being displayed to me.

So I think the template is seeing null values. This'll reflect what is being seen, because in Prosody NULL coerces to 0 or an empty string and in Unix datetime representations that 0 would translate to the "12:00 1 January 1970" date I'm seeing.

**Comment by alcinnz** [comment]
Wrote and tested that `{% debug %}` tag, which is telling me pretty much what I expected, complete with sensible non-null values.

So next I'll make sure that I'm using the right variable names.

**Comment by alcinnz** [comment]
Yes, everything matches up, and that's bad news because now I'm totally confused about what's wrong and as such I'm not sure how to fix this one.

**Comment by alcinnz** [comment]
O.K., I desperately rewrote the data bindings from SQLite to Prosody and now it magically works! I'll leave it for now and add finishing touches tomorrow before releasing a new version to the AppCenter!

Specially I'll improve the data gathering and add a few links to the history viewer so you can find that page.

**Comment by alcinnz** [comment]
I just figured why the database values weren't making their way into the template!

Previously I loaded the database row into a Gee.HashMap, but using that in combination with G.Bytes for keys as I was required some special configuration I didn't do. If I don't, no keys can be successfully looked up in it and this NULLs are returned to Prosody (done because I didn't want to deal with runtime exceptions, and because outputting an incomplete page is probably better than showing an error).

But now that I don't have a (misconfigured) hashmap that issue no longer occurs.
